src/v/s3/client.cc

```diff
@@ -0,0 +1,427 @@
+    return _client.request(std::move(header.value()))
+      .then([](http::client::response_stream_ref&& ref) {
+          // here we didn't receive any bytes from the socket and
+          // ref->is_header_done() is 'false', we need to prefetch
+          // the header first
+          return ref->prefetch_headers().then([ref = std::move(ref)]() mutable {
+              vassert(ref->is_header_done(), "Header is not received");
+              if (
+                ref->get_headers().result() != boost::beast::http::status::ok) {
+                  // Got error response, consume the response body and produce
+                  // rest api error
+                  return drain_response_stream(std::move(ref))
+                    .then([](iobuf&& res) {
+                        return parse_rest_error_response<
+                          http::client::response_stream_ref>(std::move(res));
+                    });
+              }
+              return ss::make_ready_future<http::client::response_stream_ref>(
+                std::move(ref));
+          });
+      });
+}
+
+ss::future<> client::put_object(
+  bucket_name const& name,
+  object_key const& id,
+  size_t payload_size,
+  ss::input_stream<char>&& body) {
+    auto header = _requestor.make_unsigned_put_object_request(
+      name, id, payload_size);
+    if (!header) {
+        return ss::make_exception_future<>(std::system_error(header.error()));
+    }
+    vlog(s3_log.trace, "send https request:\n{}", header);
+    return ss::do_with(
+      std::move(body),
+      [this, header = std::move(header)](ss::input_stream<char>& body) mutable {
+          return _client.request(std::move(header.value()), body)
+            .then([](const http::client::response_stream_ref& ref) {
+                return drain_response_stream(ref).then([ref](iobuf&& res) {
+                    auto status = ref->get_headers().result();
+                    if (status != boost::beast::http::status::ok) {
+                        return parse_rest_error_response<>(std::move(res));
+                    }
+                    return ss::now();
+                });
+            });
+      });
+}
+
+ss::future<client::list_bucket_result> client::list_objects_v2(
+  const bucket_name& name,
+  std::optional<object_key> prefix,
+  std::optional<object_key> start_after,
+  std::optional<size_t> max_keys) {
+    auto header = _requestor.make_list_objects_v2_request(
+      name, std::move(prefix), std::move(start_after), max_keys);
+    if (!header) {
+        return ss::make_exception_future<list_bucket_result>(
+          std::system_error(header.error()));
+    }
+    vlog(s3_log.trace, "send https request:\n{}", header);
+    return _client.request(std::move(header.value()))
+      .then([](const http::client::response_stream_ref& resp) mutable {
+          // chunked encoding is used so we don't know output size in
+          // advance
+          return ss::do_with(
+            resp->as_input_stream(),
```

Comment:
reading `client::response_stream::as_input_stream` is cool so it keeps `resp` alive properly. nice.